### ISSUE TITLE
Cannot delete file with custom role settings

### ISSUE THREAD

**Issue by Enhed** [directus/directus#15807]
### Describe the Bug

I create custom role `registered-users`, users in role have access to upload files, its ok. But deleting is forbidden:  

![image](https://user-images.githubusercontent.com/36981278/193027648-be8c8aa8-02ba-4f0c-bf1e-43bad1e1ff6c.png)

I want to users can be delete they avatars

### To Reproduce

Create role with allow upload files and delete files when ID file equals they avatar. Try to delete file with user token

### Errors Shown

403 forbidden

### What version of Directus are you using?

9.18.1

### What version of Node.js are you using?

16

### What database are you using?

Postgres 12

### What browser are you using?

Chrome

### How are you deploying Directus?

manual in docker

**Comment by emrahnazif** [comment]
Can you try : `$CURRENT_USER.avatar.id` ?

**Comment by Enhed** [comment]
> Can you try : `$CURRENT_USER.avatar.id` ?

not help

**Comment by emrahnazif** [comment]
I created a "test role".
I revoked all "Directus Files" permissions.
I added a permission for delete action for directus files with following rule: `id equals $CURRENT_USER.avatar.id` or `id equals $CURRENT_USER.avatar`
I created a "test user" with the "test role"
I added an avatar to "test user" (with admin user).
I created a token so I can test the file delete via postman

It didn't work (FORBIDDEN). 


I gave same rule `id equals $CURRENT_USER.avatar.id` or `id equals $CURRENT_USER.avatar` to read action as well. Then I got: `INTERNAL SERVER ERROR` (avatar is actually deleted)

<img width="747" alt="Screenshot 2022-09-30 at 10 26 54" src="https://user-images.githubusercontent.com/33692017/193227572-aee0ea71-c850-4421-8753-630241b091ec.png">

And also when visit the user via APP with admin user:
<img width="618" alt="Screenshot 2022-09-30 at 10 26 44" src="https://user-images.githubusercontent.com/33692017/193227721-d67f1e01-d6ca-407d-a63e-4e6978b62586.png">
I believe this is an unexpected behaviour.

**Comment by azrikahar** [comment]
> I gave same rule `id equals $CURRENT_USER.avatar.id` or `id equals $CURRENT_USER.avatar` to read action as well. Then I got: `INTERNAL SERVER ERROR` (avatar is actually deleted)
> 
> <img alt="Screenshot 2022-09-30 at 10 26 54" width="747" src="https://user-images.githubusercontent.com/33692017/193227572-aee0ea71-c850-4421-8753-630241b091ec.png">

I believe @emrahnazif's is right on the money with the read permissions requirement here, as when files are being deleted, they are being read first for the API to know which storage should the actual files be deleted, only then the actual `directus_files` rows are deleted afterwards:

https://github.com/directus/directus/blob/cbe1e7fbc69c8e9bd02254e9ed6efc35ba6468ef/api/src/services/files.ts#L282-L283

> And also when visit the user via APP with admin user: <img alt="Screenshot 2022-09-30 at 10 26 44" width="618" src="https://user-images.githubusercontent.com/33692017/193227721-d67f1e01-d6ca-407d-a63e-4e6978b62586.png"> I believe this is an unexpected behaviour.

I think the issue here is upon deleting the file, the user's `avatar` field isn't being nullified, so it's trying to load a non-existent file. Whereas if we create a usual File field on any collection, the default relationship's on delete `directus_files` trigger is set to nullify, so this issue isn't present elsewhere.

Not 100% sure, but maybe the solution would be to add the "on delete, SET NULL" foreign key triggers between `directus_users` and `directus_files` (but not sure how that interacts with custom fields?), similar to the existing ones like these:

https://github.com/directus/directus/blob/cbe1e7fbc69c8e9bd02254e9ed6efc35ba6468ef/api/src/database/migrations/20210519A-add-system-fk-triggers.ts#L15-L81

in a follow up migration.

---

@emrahnazif slightly off topic here but just want to thank you for helping out with triaging this issue! 😊

**Comment by emrahnazif** [comment]
> slightly off topic here but just want to thank you for helping out with triaging this issue!

@azrikahar I am not able to help for development of the this beautiful open source project but at least I am trying to make some test contribution to make your debugging easier. My pleasure. 

One note about the "read permissions", I only gave read access to `id`. So in my case the `storage` was not visible. Maybe that's the reason I got an unexpected result.

**Comment by rijkvanzanten** [comment]
Linear: [ENG-175](https://linear.app/directus/issue/ENG-175/cannot-delete-file-with-custom-role-settings)

**Comment by twoends** [comment]
I just had the same issue and in v10.13.2 it is necessary to allow visibility for the custom role to the following files fields:
- location
- storage
- filename_disk
- filename_download

After the deletion rule can be easily set as following:
![image](https://github.com/user-attachments/assets/fb65431c-88d6-408d-ac4b-997ca61954cc)

Since it's a configuration issue I think this can be closed @rijkvanzanten

**Comment by ComfortablyCoding** [comment]
> I just had the same issue and in v10.13.2 it is necessary to allow visibility for the custom role to the following files fields:
> 
> * location
> * storage
> * filename_disk
> * filename_download
> 
> After the deletion rule can be easily set as following: ![image](https://github.com/user-attachments/assets/fb65431c-88d6-408d-ac4b-997ca61954cc)
> 
> Since it's a configuration issue I think this can be closed [@rijkvanzanten](https://github.com/rijkvanzanten)

Agreed, closing as the original report was due to a misconfiguration. 

P.S. The additional issue highlighted in https://github.com/directus/directus/issues/15807#issuecomment-1263279383, as noted by Azri, involves a foreign key constraint related to files. This is a variation of https://github.com/directus/directus/issues/9507, and can therefore be tracked/followed there.
